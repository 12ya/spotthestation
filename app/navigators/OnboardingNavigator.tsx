import { createNativeStackNavigator } from "@react-navigation/native-stack"
import { StackScreenProps } from "@react-navigation/stack"
import React from "react"
import { Login } from "../screens/OnboardingScreen/Login"
import { NewPassword } from "../screens/OnboardingScreen/NewPassword"
import { Signup } from "../screens/OnboardingScreen/Signup"
import { SignupOTP } from "../screens/OnboardingScreen/SignupOTP"
import { SignupCompleteProfile } from "../screens/OnboardingScreen/SignupCompleteProfile"
import { Splash } from "../screens/OnboardingScreen/Splash"

export type OnboardingParamList = {
  Splash: undefined
  Login: undefined
  Signup: undefined
  SignupOTP: undefined
  SignupCompleteProfile: undefined
  NewPassword: undefined
}

export type OnboardingStackScreenProps<T extends keyof OnboardingParamList> = StackScreenProps<
  OnboardingParamList,
  T
>

const Stack = createNativeStackNavigator<OnboardingParamList>()

export function OnboardingNavigator() {
  return (
    <Stack.Navigator
      screenOptions={{ headerShown: false }}
      initialRouteName="Splash"
    >
      <Stack.Screen name="Splash" component={Splash} />
      <Stack.Screen name="Login" component={Login} />
      <Stack.Screen name="Signup" component={Signup} />
      <Stack.Screen name="SignupOTP" component={SignupOTP} />
      <Stack.Screen name="SignupCompleteProfile" component={SignupCompleteProfile} />
      <Stack.Screen name="NewPassword" component={NewPassword} />
    </Stack.Navigator>
  )
}
