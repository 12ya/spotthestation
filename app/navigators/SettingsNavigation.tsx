import { createNativeStackNavigator } from "@react-navigation/native-stack"
import { StackScreenProps } from "@react-navigation/stack"
import React from "react"
import { ContactUsScreen } from "../screens/MainScreen/SettingsScreen/ContactUsScreen"
import { TermsAndConditionsScreen } from "../screens/MainScreen/SettingsScreen/TermsAndConditionsScreen"

export type SettingsParamList = {
  TermsAndConditions: undefined
  ContactUs: undefined
}

export type SettingsStackScreenProps<T extends keyof SettingsParamList> = StackScreenProps<
  SettingsParamList,
  T
>

const Stack = createNativeStackNavigator<SettingsParamList>()

export function SettingsNavigator() {
  return (
    <Stack.Navigator
      screenOptions={{ headerShown: false }}
      initialRouteName="TermsAndConditions"
    >
      <Stack.Screen name="TermsAndConditions" component={TermsAndConditionsScreen} />
      <Stack.Screen name="ContactUs" component={ContactUsScreen} />
    </Stack.Navigator>
  )
}
