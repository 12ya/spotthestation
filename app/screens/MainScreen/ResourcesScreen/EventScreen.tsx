/* eslint-disable @typescript-eslint/restrict-template-expressions */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { useNavigation, useRoute } from "@react-navigation/native"
import { observer } from "mobx-react-lite"
import React from "react"
import { ViewStyle, TextStyle, ScrollView, Pressable, View, Image, ImageStyle } from "react-native"
import { useSafeAreaInsets } from "react-native-safe-area-context"
import { Icon, Screen, Text } from "../../../components"
import { colors, typography } from "../../../theme"
import { Tag } from "../components/Tag"

export interface EventScreenRouteProps {
  item: any
}

export const EventScreen = observer(function EventScreen() {
  const navigation = useNavigation()
  const { params: { item } } = useRoute()
  
  const topInset = useSafeAreaInsets().top

  const $headerStyleOverride: TextStyle = {
    top: topInset + 24,
  }

  return (
    <Screen
      preset="fixed" 
      contentContainerStyle={[$container, $headerStyleOverride]} 
      style={{backgroundColor: colors.palette.neutral900}} 
      statusBarStyle="light"
    >
      <ScrollView style={$scrollContainer}>
        <View>
          <Pressable onPress={() => navigation.goBack()} style={$backButton}>
            <Icon icon="caretLeft" color={colors.palette.neutral250} />
            <Text tx="resources.header" style={$backButtonText} />
          </Pressable>
          <Text text={item?.title} style={$title} />
          <View style={$tagsContainer}>
            {item?.tags?.map(tag => <Tag key={tag} title={tag} />)}
          </View>
          <Image source={{ uri: item?.image }} style={$imageContainer} resizeMode="cover" />
          <Text text={`${item?.type} Details`} style={$subtitle}/>
          <View style={$detailsContainer}>
            {item?.details?.map(({ key, value}) => (
              <View key={key} style={$detailsRow}>
                <Text text={key} style={$detailKey} />
                <Text text={value} style={$detailValue} />
              </View>
            ))}
          </View>
          <Text text={item?.subtitle} style={$subtitle}/>
          <Text text={item?.body} style={$text} />
        </View>
      </ScrollView>
    </Screen>
  )
})

const $container: ViewStyle = {
  flex: 1,
  backgroundColor: colors.backgroundDark,
  height: '100%'
}

const $detailsContainer: ViewStyle = {
  backgroundColor: colors.palette.neutral350,
  width: '100%',
  borderRadius: 10,
  marginBottom: 36,
  padding: 24
}

const $backButton: ViewStyle = {
  flexDirection: "row",
  alignItems: 'center',
  paddingBottom: 11
}

const $detailsRow: ViewStyle = {
  flexDirection: "row",
  justifyContent: 'space-between',
  paddingBottom: 16
}

const $scrollContainer: ViewStyle = { 
  paddingHorizontal: 18,
  flex: 1
}

const $title: TextStyle = {
  fontFamily: typography.primary?.normal,
  fontSize: 36,
  lineHeight: 44,
  color: colors.palette.neutral250,
  textAlign: 'left',
  paddingBottom: 12,
  textTransform: 'capitalize'
}

const $subtitle: TextStyle = {
  fontFamily: typography.primary?.normal,
  fontSize: 24,
  lineHeight: 29,
  color: colors.palette.neutral250,
  textAlign: 'left',
  paddingBottom: 16,
  textTransform: 'capitalize'
}

const $text: TextStyle = {
  fontFamily: typography.primary?.normal,
  fontSize: 18,
  lineHeight: 24,
  color: colors.palette.neutral450,
  textAlign: 'left',
  paddingBottom: 24
}

const $detailValue: TextStyle = {
  fontFamily: typography.primary?.normal,
  fontSize: 18,
  lineHeight: 21,
  color: colors.palette.neutral100,
  textAlign: 'right',
  width: '40%'
}

const $detailKey: TextStyle = {
  fontFamily: typography.primary?.normal,
  fontSize: 18,
  lineHeight: 21,
  color: colors.palette.neutral450,
  textAlign: 'left',
}

const $backButtonText: TextStyle = {
  ...$text,
  color: colors.palette.neutral250,
  paddingBottom: 0,
  paddingLeft: 5
}

const $tagsContainer: ViewStyle = {
  flexDirection: 'row',
  paddingBottom: 24
}

const $imageContainer: ImageStyle = {
  width: "100%",
  aspectRatio: 1,
  borderRadius: 12,
  marginBottom: 12
}