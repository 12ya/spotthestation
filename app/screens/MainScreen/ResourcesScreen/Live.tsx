import { StyleFn, useStyles } from "../../../utils/useStyles"
import React from "react"
import {
  ViewStyle,
  View,
  TextStyle,
  Image,
  ImageStyle,
  Text as RNText,
  Pressable,
} from "react-native"
import { Text } from "../../../components"
import { typography } from "../../../theme"
import { colors } from "../../../theme/colors"

const streamId = "P9C25Un7xaM"

export function Live({ onLink }) {
  const { $contentContainer, $title, $description, $text, $image, $flex, $link } = useStyles(styles)

  return (
    <View style={$contentContainer}>
      <Text
        accessible
        accessibilityLabel="modal title"
        accessibilityHint="modal title"
        accessibilityRole="text"
        tx="resources.liveTitle"
        style={$title}
      />

      <RNText
        accessible
        accessibilityLabel="modal text"
        accessibilityHint="modal text"
        accessibilityRole="text"
        style={$description}
      >
        <Text style={$text} tx="resources.liveDescription" />{" "}
        <Text style={[$text, $link]} tx="resources.liveLink" onPress={onLink} />
        <Text style={$text}>.</Text>
      </RNText>

      <View style={$flex} />

      <Pressable onPress={onLink}>
        <Image
          style={$image as ImageStyle}
          source={{ uri: `https://img.youtube.com/vi/${streamId}/maxresdefault.jpg` }}
        />
      </Pressable>
    </View>
  )
}

const styles: StyleFn = ({ scale, fontSizes, lineHeights }) => {
  const $contentContainer: ViewStyle = {
    width: "100%",
    paddingHorizontal: scale(18),
    paddingBottom: scale(24),
    minHeight: "100%",
  }

  const $title: TextStyle = {
    color: colors.palette.neutral250,
    width: "100%",
    fontSize: fontSizes[32],
    fontFamily: typography.primary.normal,
    lineHeight: lineHeights[44],
    paddingBottom: scale(24),
  }

  const $description: TextStyle = {
    paddingBottom: scale(36),
  }

  const $text: TextStyle = {
    color: colors.palette.neutral250,
    fontSize: fontSizes[20],
    lineHeight: lineHeights[28],
  }

  const $link: TextStyle = {
    textDecorationLine: "underline",
  }

  const $image: ViewStyle = {
    width: "100%",
    aspectRatio: 16 / 9,
  }

  const $flex: ViewStyle = {
    flex: 1,
  }

  return {
    $contentContainer,
    $title,
    $description,
    $text,
    $image,
    $flex,
    $link,
  }
}
