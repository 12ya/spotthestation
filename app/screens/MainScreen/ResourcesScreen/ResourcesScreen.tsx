/* eslint-disable @typescript-eslint/no-unsafe-argument */
/* eslint-disable react-native/no-inline-styles */
/* eslint-disable @typescript-eslint/restrict-template-expressions */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { useNavigation, useRoute } from "@react-navigation/native"
import { observer } from "mobx-react-lite"
import React, { useCallback, useEffect, useState } from "react"
import { ActivityIndicator, FlatList, ScrollView, TextStyle, View, ViewStyle } from "react-native"
import { XMLParser } from 'fast-xml-parser'
import { Accessory, Button, Icon, Screen, Text, TextField } from "../../../components"
import { api } from "../../../services/api"
import { colors, typography } from "../../../theme"
import { useSafeAreaInsetsStyle } from "../../../utils/useSafeAreaInsetsStyle"
import { ExpandContainer } from "../components/ExpandContainer"
import { FeedItem } from "../components/FeedItem"
import { FeedSearchResultItem } from "../components/FeedSearchResultItem"

const items = [{
  tags: ['history'],
  title: "Expedition 68 Welcomes Crew-6 Members Aboard Station",
  date: "2023-03-03T00:00:00.000000",
  image: "https://blogs.nasa.gov/spacestation/wp-content/uploads/sites/240/2023/03/blog_crew_crew_greeting.jpg",
  type: "event",
  link: 'https://blogs.nasa.gov/spacestation/2023/03/03/expedition-68-welcomes-crew-6-members-aboard-station/'
}, {
  tags: ['history'],
  title: "International Space Station Facts and Figures",
  date: "2023-01-04T00:00:00.000000",
  image: "https://www.nasa.gov/sites/default/files/thumbnails/image/iss-blowout_updated_view-web-version-010323.png",
  type: "event",
  link: 'https://www.nasa.gov/feature/facts-and-figures'
}]

const suggestions = ['NASA Upcoming Missons', 'Interviews of the Week', 'ARTEMIS II']

export interface ResourcesScreenRouteProps {
  toggleBottomTabs: (value: boolean) => void
}

export const Resources = observer(function HomeScreen() {
  const navigation = useNavigation()
  const route: ResourcesScreenRouteProps = useRoute().params as ResourcesScreenRouteProps
  const $topInset = useSafeAreaInsetsStyle(["top", "bottom"], "padding")
  const [isSearch, setIsSearch] = useState(false)
  const [isLoading, setIsLoading] = useState(false)
  const [searchQuery, setSearchQuery] = useState('')
  const [type, setType] = useState('about')
  const [news, setNews] = useState([])
  const [page, setPage] = useState(1)

  const fetchData = useCallback((paged = 1) => {
    setIsLoading(true)
    api.getFeed(+paged).then(res => {
      const parser = new XMLParser()
      const jObj = parser.parse(res.places)
      setNews([...news, ...jObj.rss.channel.item])
      setIsLoading(false)
    }).catch(e => {
      setIsLoading(false)
      console.log(e)
    })
  }, [news])

  useEffect(() => {
    fetchData()
  }, [])

  const renderFooter = () => {
    if (isLoading) {
      return (
        <View style={{ paddingVertical: 20 }}>
          <ActivityIndicator size="large" />
        </View>
      )
    } else {
      return null
    }
  }

  const renderStatic = () => {
    return (
      <ScrollView 
        accessible
        accessibilityLabel="recent results"
        accessibilityHint="recent results"
        accessibilityRole="scrollbar"
        style={$scrollContainer}
      >
        <View style={[$scrollContainer, $bodyContainer]}>
          {items.map(item => <FeedItem
                      key={item.title}
                      onPress={() => navigation.navigate('ResourcesScreens' as never, { screen: 'Events', item: item.link } as never)}
                      tags={item.tags} 
                      title={item.title}
                      date={(new Date(item.date)).toISOString()}
                      image={item.image}
                    />)}
        </View>
      </ScrollView>
    )
  }

  const renderSearch = useCallback(() => {
    if (searchQuery) {
      return <ExpandContainer title="resources.searchResults" itemsCount={items.length} expandble={false}>
        <FlatList 
          data={[...news].filter(item => item.title.toLowerCase().includes(searchQuery.toLowerCase()))} 
          renderItem={({item}) => <FeedSearchResultItem
            key={item.title}
            onPress={() => navigation.navigate('ResourcesScreens' as never, { screen: 'Events', item: item.guid } as never)}
            // tags={item.tags}
            title={item.title}
            type={item.type}
            image={/<img.*?src="([^"]*)"/.exec(item["content:encoded"]) ? /<img.*?src="([^"]*)"/.exec(item["content:encoded"])[1] : ''}
          />}
          numColumns={2}
          columnWrapperStyle={{ justifyContent: 'space-between' }}
          contentContainerStyle={{ paddingBottom: 200 }}
        />
      </ExpandContainer>
    } else {
      return (
        <ScrollView 
          accessible
          accessibilityLabel="recent results"
          accessibilityHint="recent results"
          accessibilityRole="scrollbar"
          style={$scrollContainer}
        >
          <View style={[$scrollContainer, $bodyContainer]}>
            <ExpandContainer title="resources.suggestions" expandble={false}>
              {suggestions.map(suggestion => (<View key={suggestion} style={$suggestionContainer}>
                <Icon icon='search' size={28} />
                <Text
                  accessible
                  accessibilityLabel="suggestion"
                  accessibilityHint={suggestion}
                  accessibilityRole="text"
                  text={suggestion} 
                  style={$suggestion}
                />
              </View>))}
            </ExpandContainer>
          </View>
        </ScrollView>
      )
    }
  }, [searchQuery, news])

  const renderBody = useCallback(() => {
    if (isSearch) {
      route.toggleBottomTabs(false)
      return renderSearch()
    } else {
      route.toggleBottomTabs(true)
      return (
        <FlatList 
          data={news} 
          renderItem={({item}) => <FeedItem
                      key={item.title}
                      onPress={() => navigation.navigate('ResourcesScreens' as never, { screen: 'Events', item: item.guid } as never)} 
                      // tags={item.tags} 
                      title={item.title}
                      date={(new Date(item.pubDate)).toISOString()}
                      image={/<img.*?src="([^"]*)"/.exec(item["content:encoded"]) ? /<img.*?src="([^"]*)"/.exec(item["content:encoded"])[1] : ''}
                    />}
          numColumns={2}
          columnWrapperStyle={{ justifyContent: 'space-between'}}
          onEndReached={() => {
            setPage(page + 1)
            fetchData(page + 1)
          }}
          onEndReachedThreshold={0.5}
          ListFooterComponent={renderFooter}
        />
      )
    }
  }, [isSearch, renderSearch, news])

  return (
    <Screen preset="fixed" contentContainerStyle={$container} style={[$topInset, {backgroundColor: colors.palette.neutral900}]} statusBarStyle="light">
      <View style={$headerContainer}>
        {isSearch ? (<View style={{ flex: 1, marginRight: 18 }}>
          <TextField
            accessible
            accessibilityLabel="search"
            accessibilityHint="type to search events"
            accessibilityRole="search"
            value={searchQuery}
            inputWrapperStyle={$searchField}
            placeholderTx="resources.searchPlaceholder"
            onChangeText={setSearchQuery}
            renderLeftAccessory={({ style }) => (
              <Accessory 
                icon="search"
                color={colors.palette.neutral450}
                style={style}
              />
            )}
          />
        </View>)
         : <Text 
            accessible
            accessibilityLabel="header"
            accessibilityHint="header"
            accessibilityRole="text"
            tx="resources.header" 
            style={$header} 
          />}
        <Icon icon={isSearch ? 'x' : 'search'} size={24} containerStyle={[$searchButton, isSearch && $xButton]} onPress={() => setIsSearch(!isSearch)} />
      </View>
      <View style={$horizontalScrollContainer}>
        <ScrollView horizontal style={$horizontalScrollContainer}>
          {['about', 'news'].map(item => <Button
            key={item}
            accessible
            accessibilityLabel={`${item} button`}
            accessibilityHint={`show ${item} view`}
            text={item}
            style={[$button, item === type && $active]}
            textStyle={$buttonText}
            pressedStyle={$button}
            onPress={() => setType(item)}
          />)}
        </ScrollView>
      </View>
      {type === 'news' ? renderBody() : renderStatic()}
    </Screen>
  )
})

const $container: ViewStyle = {
  flex: 1,
  backgroundColor: colors.backgroundDark,
  paddingHorizontal: 18
}

const $headerContainer: ViewStyle = {
  flexDirection: 'row',
  justifyContent: 'space-between',
  alignItems: 'center'
}

const $searchButton: ViewStyle = {
  width: 42, 
  height: 42, 
  borderRadius: 24, 
  backgroundColor: colors.palette.neutral350,
  alignItems: 'center',
  justifyContent: 'center'
}

const $xButton: ViewStyle = {
  ...$searchButton,
  width: 56, 
  height: 56, 
  borderRadius: 36,
  backgroundColor: colors.palette.neutral550,
}

const $bodyContainer: ViewStyle = {
  justifyContent: "space-between",
  flexDirection: 'row',
  flexWrap: "wrap",
  paddingBottom: 200
}

const $suggestionContainer: ViewStyle = {
  flexDirection: 'row',
  alignItems: 'center',
  paddingBottom: 24
}

const $header: TextStyle = {
  fontFamily: typography.primary.normal,
  fontSize: 36,
  lineHeight: 44,
  color: colors.palette.neutral250,
}

const $suggestion: TextStyle = {
  fontFamily: typography.primary.normal,
  fontSize: 18,
  lineHeight: 21,
  color: colors.palette.neutral250,
  paddingLeft: 12
}

const $scrollContainer: ViewStyle = { 
  flex: 1,
}

const $horizontalScrollContainer: ViewStyle = { 
  width: '100%',
  height: 80,
  marginTop: 10
}

const $searchField: ViewStyle = {
  borderWidth: 1.5,
  borderColor: "transparent",
  borderRadius: 28,
  height: 56,
  backgroundColor: colors.palette.neutral550,
  overflow: "hidden",
}

const $button: ViewStyle = {
  width: '40%',
  height: 46,
  minHeight: 40,
  backgroundColor: 'transparent',
  borderRadius: 28,
  borderWidth: 0,
  marginRight: 24,
  paddingHorizontal: 20
}

const $active: ViewStyle = {
  backgroundColor: colors.palette.neutral550
}

const $buttonText: TextStyle = {
  color: colors.palette.neutral100,
  fontSize: 18,
  fontFamily: typography.primary.medium,
  lineHeight: 22,
  textTransform: 'capitalize'
}
