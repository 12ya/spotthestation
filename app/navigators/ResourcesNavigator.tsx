import { useRoute } from "@react-navigation/native"
import { createNativeStackNavigator } from "@react-navigation/native-stack"
import { StackScreenProps } from "@react-navigation/stack"
import React from "react"
import { EventScreen, EventScreenRouteProps } from "../screens/MainScreen/ResourcesScreen/EventScreen"

export type ResourcesParamList = {
  Event: EventScreenRouteProps
}

export type ResourcesStackScreenProps<T extends keyof ResourcesParamList> = StackScreenProps<
ResourcesParamList,
  T
>

const Stack = createNativeStackNavigator<ResourcesParamList>()

export function ResourcesNavigator() {
  const { params } = useRoute()
  return (
    <Stack.Navigator
      screenOptions={{ headerShown: false }}
      initialRouteName="Event"
    >
      <Stack.Screen name="Event" component={EventScreen} initialParams={params}/>
    </Stack.Navigator>
  )
}
