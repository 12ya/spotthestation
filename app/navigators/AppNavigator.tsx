/**
 * The app navigator (formerly "AppNavigator" and "MainNavigator") is used for the primary
 * navigation flows of your app.
 * Generally speaking, it will contain an auth flow (registration, login, forgot password)
 * and a "main" flow which the user will use once logged in.
 */
import {
  DarkTheme,
  DefaultTheme,
  NavigationContainer,
  NavigatorScreenParams,
} from "@react-navigation/native"
import { createNativeStackNavigator } from "@react-navigation/native-stack"
import { StackScreenProps } from "@react-navigation/stack"
import { observer } from "mobx-react-lite"
import React, { useEffect, useRef } from "react"
import { AppState, Platform, useColorScheme } from "react-native"
import * as storage from "../utils/storage"
import Config from "../config"
import { TabNavigator, TabParamList } from "./TabNavigator"
import { OnboardingNavigator, OnboardingParamList } from "./OnboardingNavigator"
import { navigationRef, skipOnboarding, useBackButtonHandler } from "./navigationUtilities"
import { SettingsNavigator, SettingsParamList } from "./SettingsNavigation"
import { ResourcesNavigator, ResourcesParamList } from "./ResourcesNavigator"
import Snackbar from "react-native-snackbar"
import { translate } from "../i18n"
import * as notifications from "../utils/notifications"
import notifee, { EventType } from "@notifee/react-native"

/**
 * This type allows TypeScript to know what routes are defined in this navigator
 * as well as what properties (if any) they might take when navigating to them.
 *
 * If no params are allowed, pass through `undefined`. Generally speaking, we
 * recommend using your MobX-State-Tree store(s) to keep application state
 * rather than passing state through navigation params.
 *
 * For more information, see this documentation:
 *   https://reactnavigation.org/docs/params/
 *   https://reactnavigation.org/docs/typescript#type-checking-the-navigator
 *   https://reactnavigation.org/docs/typescript/#organizing-types
 */
export type AppStackParamList = {
  Main: NavigatorScreenParams<TabParamList>
  Onboarding: NavigatorScreenParams<OnboardingParamList>
  SettingsScreens: NavigatorScreenParams<SettingsParamList>
  ResourcesScreens: NavigatorScreenParams<ResourcesParamList>
}

/**
 * This is a list of all the route names that will exit the app if the back button
 * is pressed while in that screen. Only affects Android.
 */
const exitRoutes = Config.exitRoutes

export type AppStackScreenProps<T extends keyof AppStackParamList> = StackScreenProps<
  AppStackParamList,
  T
>

// Documentation: https://reactnavigation.org/docs/stack-navigator/
const Stack = createNativeStackNavigator<AppStackParamList>()

const AppStack = observer(function AppStack() {
  const prevAppState = useRef<string>()

  useEffect(() => {
    Promise.all([storage.load("isSettingsCompleted"), notifications.hasInitialNotification()])
      .then(([isSettingsCompleted, hasInitialNotification]) => {
        if (isSettingsCompleted) skipOnboarding()
        if (hasInitialNotification) {
          navigationRef.navigate(
            "Main" as never,
            { screen: "ISSView", params: { info: true } } as never,
          )
        }
      })
      .catch((err) =>
        Snackbar.show({
          text: err as string,
          duration: Snackbar.LENGTH_LONG,
          action: {
            text: translate("snackBar.dismiss"),
            textColor: "red",
            onPress: () => {
              Snackbar.dismiss()
            },
          },
        }),
      )

    if (Platform.OS === "android") {
      const subscription = AppState.addEventListener("change", (nextAppState) => {
        if (
          nextAppState === "active" &&
          prevAppState.current &&
          prevAppState.current !== nextAppState
        )
          notifications
            .hasInitialNotification()
            .then((hasInitialNotification) => {
              if (hasInitialNotification)
                navigationRef.navigate(
                  "Main" as never,
                  { screen: "ISSView", params: { info: true } } as never,
                )
            })
            .catch((e) => console.error(e))
        prevAppState.current = nextAppState
      })

      return () => {
        subscription.remove()
      }
    } else {
      return notifee.onForegroundEvent(({ type }) => {
        if (type === EventType.PRESS) {
          navigationRef.navigate(
            "Main" as never,
            { screen: "ISSView", params: { info: true } } as never,
          )
        }
      })
    }
  }, [])

  return (
    <Stack.Navigator screenOptions={{ headerShown: false }} initialRouteName={"Onboarding"}>
      <Stack.Screen name="Onboarding" component={OnboardingNavigator} />
      <Stack.Screen name="Main" component={TabNavigator} />
      <Stack.Screen name="SettingsScreens" component={SettingsNavigator} />
      <Stack.Screen name="ResourcesScreens" component={ResourcesNavigator} />
    </Stack.Navigator>
  )
})

interface NavigationProps extends Partial<React.ComponentProps<typeof NavigationContainer>> {}

export const AppNavigator = observer(function AppNavigator(props: NavigationProps) {
  const colorScheme = useColorScheme()

  useBackButtonHandler((routeName) => exitRoutes.includes(routeName))

  return (
    <NavigationContainer
      ref={navigationRef}
      theme={colorScheme === "dark" ? DarkTheme : DefaultTheme}
      {...props}
    >
      <AppStack />
    </NavigationContainer>
  )
})
