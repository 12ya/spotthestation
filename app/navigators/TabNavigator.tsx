/* eslint-disable react-native/no-color-literals */
/* eslint-disable react-native/no-inline-styles */
import { BottomTabScreenProps, createBottomTabNavigator } from "@react-navigation/bottom-tabs"
import { CompositeScreenProps } from "@react-navigation/native"
import React, { useState } from "react"
import { TextStyle, ViewStyle } from "react-native"
import { useSafeAreaInsets } from "react-native-safe-area-context"
import { Icon, Text } from "../components"
import { ISSNowScreen, HomeScreen, ISSNowScreenRouteProps, SkyViewScreenRouteProps, SkyViewScreen, Resources } from "../screens"
import { colors, spacing, typography } from "../theme"
import { AppStackParamList, AppStackScreenProps } from "./AppNavigator"

export type TabParamList = {
  Home: undefined
  SkyView: SkyViewScreenRouteProps
  ISSNow: ISSNowScreenRouteProps
  Resources: undefined
  Settings: undefined
}

/**
 * Helper for automatically generating navigation prop types for each route.
 *
 * More info: https://reactnavigation.org/docs/typescript/#organizing-types
 */
export type TabScreenProps<T extends keyof TabParamList> = CompositeScreenProps<
  BottomTabScreenProps<TabParamList, T>,
  AppStackScreenProps<keyof AppStackParamList>
>

const Tab = createBottomTabNavigator<TabParamList>()

export function TabNavigator() {
  const { bottom } = useSafeAreaInsets()
  const [isTabsVisible, setIsTabsVisible] = useState(true)
  
  return (
    <Tab.Navigator
      initialRouteName="Home"
      screenOptions={{
        headerShown: false,
        tabBarActiveTintColor: colors.palette.neutral100,
        tabBarStyle: [$tabBar, { height: bottom + 70, display: isTabsVisible ? "flex" : "none" }],
        tabBarLabelStyle: $tabBarLabel,
        tabBarItemStyle: $tabBarItem,
      }}
    >
      <Tab.Screen
        name="Home"
        component={HomeScreen}
        options={{
          tabBarLabel: ({ focused, color }) => (
            <Text tx="tabNavigator.homeTab" style={{ color: focused ? color : "transparent" }} />
          ),
          tabBarIcon: ({ color, size }) => (
            <Icon icon="home" color={color} size={size} />
          ),
        }}
      />

      <Tab.Screen
        name="SkyView"
        component={SkyViewScreen}
        initialParams={{
          toggleBottomTabs: setIsTabsVisible
        }}
        options={{
          tabBarLabel: ({ focused, color }) => (
            <Text tx="tabNavigator.skyViewTab" style={{ color: focused ? color : "transparent" }} />
          ),
          tabBarIcon: ({ color, size }) => (
            <Icon icon="globe" color={color} size={size} />
          ),
        }}
      />

      <Tab.Screen
        name="ISSNow"
        component={ISSNowScreen}
        initialParams={{
          toggleBottomTabs: setIsTabsVisible
        }}
        options={{
          tabBarLabel: ({ focused, color }) => (
            <Text tx="tabNavigator.issNowTab" style={{ color: focused ? color : "transparent" }} />
          ),
          tabBarIcon: ({ color, size }) => (
            <Icon icon="tv" color={ color } size={size} />
          ),
        }}
      />

      <Tab.Screen
        name="Resources"
        component={Resources}
        options={{
          tabBarLabel: ({ focused, color }) => (
            <Text tx="tabNavigator.resourcesTab" style={{ color: focused ? color : "transparent" }} ellipsizeMode="tail" numberOfLines={1} />
          ),
          tabBarIcon: ({ color, size }) => (
            <Icon icon="book" color={ color } size={size} />
          ),
        }}
      />

      <Tab.Screen
        name="Settings"
        component={HomeScreen}
        options={{
          tabBarLabel: ({ focused, color }) => (
            <Text tx="tabNavigator.settingsTab" style={{ color: focused ? color : "transparent" }} />
          ),
          tabBarIcon: ({ color, size }) => (
            <Icon icon="settings" color={ color } size={size} />
          ),
        }}
      />
    </Tab.Navigator>
  )
}

const $tabBar: ViewStyle = {
  backgroundColor: colors.backgroundDark,
  borderTopColor: colors.transparent,
}

const $tabBarItem: ViewStyle = {
  paddingTop: spacing.small,
}

const $tabBarLabel: TextStyle = {
  fontSize: 12,
  fontFamily: typography.primary.medium,
  lineHeight: 16,
  flex: 1,
}
