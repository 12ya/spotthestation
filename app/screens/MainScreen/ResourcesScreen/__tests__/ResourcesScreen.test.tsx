import React from "react"
import { Resources } from "../ResourcesScreen"
import { NavigationContainer } from "@react-navigation/native"
import { act, render } from "@testing-library/react-native"
import { TabNavigatorContext } from "../../../../navigators"

jest.mock("@react-navigation/native", () => ({
  ...Object.assign({}, jest.requireActual("@react-navigation/native")),
  useRoute: () => ({
    params: {
      id: 1,
    },
  }),
}))
it("renders correctly", async () => {
  const component = render(
    <TabNavigatorContext.Provider
      value={{ toggleBottomTabs: jest.fn(), toggleIsLandscape: jest.fn() }}
    >
      <NavigationContainer>
        <Resources />
      </NavigationContainer>
    </TabNavigatorContext.Provider>,
  )

  await act(async () => {
    await new Promise((resolve) => setTimeout(resolve, 100))
  })

  expect(component.toJSON()).toMatchSnapshot()
})
