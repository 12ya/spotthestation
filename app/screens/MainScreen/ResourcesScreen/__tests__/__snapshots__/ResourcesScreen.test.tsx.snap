// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <div
      style={
        Object {
          "color": "#DEDDDE",
          "fontFamily": "Helvetica Neue",
          "fontSize": 52,
          "lineHeight": 63,
        }
      }
    />
    <div />
  </View>
  <View
    style={
      Object {
        "height": 115,
        "marginTop": 15,
      }
    }
  >
    <RCTScrollView
      horizontal={true}
      style={
        Object {
          "height": 115,
          "marginTop": 15,
        }
      }
    >
      <View>
        <div />
        <div />
        <div />
      </View>
    </RCTScrollView>
  </View>
  <RCTScrollView
    ListFooterComponent={[Function]}
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
      />
    </View>
  </RCTScrollView>
</div>
`;
